File: csharp/01_variables.cs
// ============================================================
// Topic:  Variables
// Summary: var, explicit types, const, naming, static typing.
// Run:     dotnet script csharp/01_variables.cs
// ============================================================

using System;

// ----- 1. var vs explicit type (prefer var when the type is obvious) -----
var name = "Vetri";       // inferred as string   (Python: name = "Vetri")
int age = 22;               // explicit type        (TS: let age: number = 22)
Console.WriteLine($"{name} {age}"); // Output: Vetri 22

// ----- 2. const vs let (const is compile-time constant) -----
const double PI = 3.14159;  // cannot reassign   (TS: const PI = 3.14159)
int score = 10;
score = 20;                 // OK — regular variable is mutable
Console.WriteLine(score);   // Output: 20
// PI = 3.0;  // ❌ ERROR: const cannot be reassigned (Python has no real const)

// ----- 3. Multiple assignment via tuples -----
(int x, int y, int z) = (1, 2, 3);
int a = 0, b = 0, c = 0;    // all initialized to 0
Console.WriteLine($"{x} {y} {z} {a} {b} {c}"); // Output: 1 2 3 0 0 0

// ----- 4. Swapping (tuple deconstruction) -----
(x, y) = (y, x);            // (Python: x, y = y, x)   (TS: [x, y] = [y, x])
Console.WriteLine($"{x} {y}"); // Output: 2 1

// ----- 5. Type annotations are enforced at compile time -----
string username = "abc";
int count = 5;
bool active = true;
Console.WriteLine($"{username} {count} {active}"); // Output: abc 5 True

// ----- 6. Naming: camelCase for locals (TS convention; Python uses snake_case) -----
double totalPrice = 99.5;
Console.WriteLine(totalPrice); // Output: 99.5

// ----- 7. Static typing (no silent type change on rebind) -----
object thing = 5;
thing = "now a string";     // allowed via object   (Python allows this on any variable)
Console.WriteLine(thing);   // Output: now a string
// int n = 5; n = "text";   // ❌ ERROR: cannot assign string to int (TS would also error)

// Cross-language note:
//   C#:      var name = "Vetri";  or  string name = "Vetri";
//   TS:      let name = "Vetri";  or  const name = "Vetri";
//   Python:  name = "Vetri"        // no keyword, no real const
//   C# is statically typed like TS; Python rebinds freely (unless you use type hints).

File: java/01_variables.java
// ============================================================
// Topic:  Variables
// Summary: var, explicit types, final, naming, static typing.
// Run:     java java/01_variables.java   (Java 11+ single-file launch)
// ============================================================

public class Variables {
    public static void main(String[] args) {
        // ----- 1. var vs explicit type (var since Java 10) -----
        var name = "Vetri";        // inferred String   (Python: name = "Vetri")
        int age = 22;              // explicit type      (TS: let age: number = 22)
        System.out.println(name + " " + age); // Output: Vetri 22

        // ----- 2. final = constant (cannot reassign) -----
        final double PI = 3.14159;
        int score = 10;
        score = 20;                // OK — non-final is mutable
        System.out.println(PI + " " + score); // Output: 3.14159 20
        // PI = 3.0;  // ERROR: cannot assign a value to final variable PI

        // ----- 3. Multiple declaration on one line -----
        int x = 1, y = 2, z = 3;
        System.out.println(x + " " + y + " " + z); // Output: 1 2 3

        // ----- 4. Swapping (no tuple — use a temp variable) -----
        int a = 1, b = 2;
        int temp = a; a = b; b = temp;  // (Python: a, b = b, a)
        System.out.println(a + " " + b); // Output: 2 1

        // ----- 5. Type is fixed once declared (static typing) -----
        String username = "abc";
        int count = 5;
        boolean active = true;
        System.out.println(username + " " + count + " " + active); // Output: abc 5 true

        // ----- 6. Naming: camelCase for variables (like TS; Python uses snake_case) -----
        double totalPrice = 99.5;
        System.out.println(totalPrice); // Output: 99.5

        // Cross-language note:
        //   Java `var` infers type but it stays static (like C# var, TS let, Rust let).
        //   `final` ≈ const (C#/TS); Python has no true const, Rust uses `const`/`let`.
        //   No tuple swap — Java needs a temp var (Python/Rust/C#: (a,b)=(b,a)).
        //   camelCase variables; ClassNames are PascalCase.
    }
}
